chains/humanode/server/src/lib.rs

```diff
@@ -0,0 +1,503 @@
+        account_info.map_or_else(
+            || {
+                Ok(AccountInfo::<u32, AccountData<u128>> {
+                    nonce: 0,
+                    consumers: 0,
+                    providers: 0,
+                    sufficients: 0,
+                    data: AccountData { free: 0, reserved: 0, misc_frozen: 0, fee_frozen: 0 },
+                })
+            },
+            |account_info| {
+                <AccountInfo<u32, AccountData<u128>>>::decode(&mut account_info.encoded())
+                    .map_err(|_| anyhow::anyhow!("invalid format"))
+            },
+        )
+    }
+}
+
+#[async_trait::async_trait]
+impl BlockchainClient for HumanodeClient {
+    type MetadataParams = HumanodeMetadataParams;
+    type Metadata = HumanodeMetadata;
+    type EventStream<'a> = <MaybeWsEthereumClient as BlockchainClient>::EventStream<'a>;
+    type Call = EthQuery;
+    type CallResult = EthQueryResult;
+
+    type AtBlock = PartialBlockIdentifier;
+    type BlockIdentifier = BlockIdentifier;
+
+    type Query = rosetta_config_ethereum::Query;
+    type Transaction = rosetta_config_ethereum::SignedTransaction;
+    type Subscription = <MaybeWsEthereumClient as BlockchainClient>::Subscription;
+    type Event = <MaybeWsEthereumClient as BlockchainClient>::Event;
+    type SubmitResult = <MaybeWsEthereumClient as BlockchainClient>::SubmitResult;
+
+    async fn query(
+        &self,
+        query: Self::Query,
+    ) -> Result<<Self::Query as rosetta_core::traits::Query>::Result> {
+        self.client.query(query).await
+    }
+
+    fn config(&self) -> &BlockchainConfig {
+        self.client.config()
+    }
+
+    fn genesis_block(&self) -> Self::BlockIdentifier {
+        self.client.genesis_block()
+    }
+
+    async fn current_block(&self) -> Result<Self::BlockIdentifier> {
+        self.client.current_block().await
+    }
+
+    async fn finalized_block(&self) -> Result<Self::BlockIdentifier> {
+        self.client.finalized_block().await
+    }
+
+    async fn balance(&self, address: &Address, block: &Self::AtBlock) -> Result<u128> {
+        let balance = match address.format() {
+            AddressFormat::Ss58(_) => {
+                let account_info = self.account_info(address, Some(block)).await?;
+                account_info.data.free
+            },
+            AddressFormat::Eip55 => {
+                // Frontier `eth_getBalance` returns the reducible_balance instead the free balance:
+                // https://github.com/paritytech/frontier/blob/polkadot-v0.9.43/frame/evm/src/lib.rs#L853-L855
+                // using substrate to get the free balance
+                let address = address
+                    .evm_to_ss58(Ss58AddressFormat::custom(42))
+                    .map_err(|err| anyhow::anyhow!("{}", err))?;
+                let account_info = self.account_info(&address, Some(block)).await?;
+                account_info.data.free
```

Comment:
Humanode version returns the exact free balance, we apply own patch to meet general eth logic - https://github.com/humanode-network/frontier/commit/a06c5f0d5db777efc9374480a23ac4ea0053f395.

Reply:
Also, note that for the EVM integration you'd need to use the JSON-RPC API to interact with the network. Alternatively, you can use the `evm_system` pallet:
<img width="988" alt="image" src="https://github.com/Analog-Labs/chain-connectors/assets/265507/13af3f0a-e8ea-4bdc-aaf3-a23f8d8c3a8b">
It is not possible though to submit an EVM transaction signing it with Substrate account signature.